Cannot parse zero-sized slice patterns

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn foo() {
    let [] = [0; 0];
}
```

I expected to see this happen: no errors

Instead, this happened: 
```rust
test.rs:2:10: error: unexpected token ‘]’ in pattern
    2 |     let [] = [0; 0];
      |          ^
test.rs:2:10: error: failed to parse initial pattern in slice pattern
test.rs:2:10: error: failed to parse pattern in let statement
test.rs:2:5: error: failed to parse statement or expression without block in block expression
    2 |     let [] = [0; 0];
      |     ^~~
test.rs:2:18: error: unrecognised token ‘integer literal’ for start of item
    2 |     let [] = [0; 0];
      |                  ^
test.rs:2:18: error: failed to parse item in crate

Analyzing compilation unit

Time variable                                   usr           sys          wall           GGC
 TOTAL                              :   0.00          0.00          0.00          145k
Extra diagnostic checks enabled; compiler may run slowly.
Configure with --enable-checking=release to disable checks.
```

### Meta

0d5fe7b9aee

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -11133,7 +11133,15 @@ std::unique_ptr<AST::SlicePattern>
 Parser<ManagedTokenSource>::parse_slice_pattern ()
 {
   Location square_locus = lexer.peek_token ()->get_locus ();
+  std::vector<std::unique_ptr<AST::Pattern>> patterns;
   skip_token (LEFT_SQUARE);
+
+  if (lexer.peek_token ()->get_id () == RIGHT_SQUARE)
+    {
+      skip_token (RIGHT_SQUARE);
+      return std::unique_ptr<AST::SlicePattern> (
+	new AST::SlicePattern (std::move (patterns), square_locus));
+    }
 
   // parse initial pattern (required)
   std::unique_ptr<AST::Pattern> initial_pattern = parse_pattern ();
@@ -11146,7 +11154,6 @@ Parser<ManagedTokenSource>::parse_slice_pattern ()
       return nullptr;
     }
 
-  std::vector<std::unique_ptr<AST::Pattern>> patterns;
   patterns.push_back (std::move (initial_pattern));
 
   const_TokenPtr t = lexer.peek_token ();

--- a/gcc/testsuite/rust/compile/zero_sized_slice.rs
+++ b/gcc/testsuite/rust/compile/zero_sized_slice.rs
@@ -1,0 +1,5 @@
+// { dg-options "-fsyntax-only" }
+
+fn foo() {
+    let [] = [0; 0];
+}

